### DISCUSSION TITLE
DKI model fit shape broadcast error

### DISCUSSION THREAD

**Discussion by smeisler** [dipy/dipy#2619] [Reconstruction] [answered]
## Description
Hello,

I am fitting DKI models on HCP-preprocessed data, and am running into an error when fitting the DKI model, as described below.

### Code example to reproduce
```
# Import packages
import dipy
import os
import numpy as np
import dipy.reconst.dki as dki
from dipy.io.image import load_nifti
from dipy.core.gradients import gradient_table
from dipy.io import read_bvals_bvecs
from dipy.io.utils import nifti1_symmat
from scipy.ndimage.filters import gaussian_filter

# Get paths to data and load them
dwi_path = os.path.join(work_dir,'raw','data.nii.gz')
dwi_data, dwi_affine, dwi_img = load_nifti(dwi_path, return_img=True)
mask_path = os.path.join(work_dir,'raw','nodif_brain_mask.nii.gz')
maskdata, maskaffine, mask = load_nifti(mask_path, return_img=True)
bvec_path = os.path.join(work_dir,'raw','bvecs_corr')
bval_path = os.path.join(work_dir,'raw','bvals_corr')
bvals, bvecs = read_bvals_bvecs(bval_path, bvec_path)
gtab = gradient_table(bvals, bvecs)

# Fit DKI Model
dkimodel = dki.DiffusionKurtosisModel(gtab)
dkifit = dkimodel.fit(dwi_data, mask=mask) # <---- WHERE ERROR OCCURS
```

### Error
```
Traceback (most recent call last):
  File "/nese/mit/group/gablab/data/HCP/HCP_style/code/dwi_postproc/dki.py", line 33, in <module>
    dkifit = dkimodel.fit(dwi_data, mask=mask)
  File "/om2/user/smeisler/anaconda3/envs/pyafq/lib/python3.9/site-packages/dipy/reconst/dki.py", line 1638, in fit
    dki_params[mask, :] = params_in_mask
ValueError: shape mismatch: value array of shape (3658350,27) could not be broadcast to indexing result of shape (1,145,174,145,27)

```

## Way to reproduce
- Operating system and version: `Linux-3.10.0-1062.el7.x86_64-x86_64-with-glibc2.17`. CentOs7 HPC.
- Python version 3.9.12
- dipy version 1.5.0
- dependency versions
    * numpy 1.21.4
    * scipy 1.7.1
    * nibabel 3.2.1
    * h5py 3.5.0
    * cvxpy not installed
    * fury not installed

**Comment by smeisler** [comment]
Update, I used the Otsu masking method like in the tutorial, and it finished okay, but it is unclear why my predefined brain mask did not work.

**Comment by arokem** [comment]
Hey Steven - thanks for reporting this. What is the shape of your mask volume (i.e., what is `maskdata.shape`)? I'm suspecting an extra singleton dimension might have been added somewhere, or something of that sort.

**Comment by arokem** [comment, marked as answer]
Oh, another option: what happens if you run: 

```
dkifit = dkimodel.fit(dwi_data, mask=maskdata)
```

instead of the code you ran? I think that last output is a nibabel `Nifti1Image` object, and that might be the cause of the error.

**Comment by smeisler** [comment]
That worked!
